community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/idp/cartesianProductsOrValueJoins.scala

```diff
@@ -49,28 +49,56 @@ case object cartesianProductsOrValueJoins extends JoinDisconnectedQueryGraphComp
 
     assert(plans.size > 1, "Can't build cartesian product with less than two input plans")
 
-    // Map of newly found plans to originating LHS and RHS plans to join
-    val connectedPlans: Map[PlannedComponent, (PlannedComponent, PlannedComponent)] = {
-      val joins = produceJoinVariations(plans, qg)
+    /*
+    To connect disconnected query parts, we have a couple of different ways. First we check if there are any joins that
+    we could do. Joins are equal or better than cartesian products, so we always go for the joins when possible.
 
-      if (joins.nonEmpty)
-        joins
-      else
-        produceCartesianProducts(plans, qg)
+    Next we can do an exhaustive search for how to combine the remaining query parts together. In-between each step
+    we check if any joins have been made available. This makes for better plans, but is exponentially expensive.
+
+    So, when we have too many plans to combine, we fall back to the naive way of just building a left deep tree with
+    all query parts cross joined together.
+     */
+    val joins = produceJoinVariations(plans, qg)
+
+    if (joins.nonEmpty) {
+      pickTheBest(plans, kit, joins)
+    } else if (plans.size < 8) {
+      val cartesianProducts = produceCartesianProducts(plans, qg)
+      pickTheBest(plans, kit, cartesianProducts)
+    }
+    else {
+      planLotsOfCartesianProducts(plans, qg)
     }
+  }
 
-    val bestPlan = kit.pickBest(connectedPlans.map(_._1.plan)).get
-    val bestQG: QueryGraph = connectedPlans.collectFirst {
+  private def pickTheBest(plans: Set[PlannedComponent], kit: QueryPlannerKit, joins: Map[PlannedComponent, (PlannedComponent, PlannedComponent)]): Set[PlannedComponent] = {
+    val bestPlan = kit.pickBest(joins.map(_._1.plan)).get
+    val bestQG: QueryGraph = joins.collectFirst {
       case (PlannedComponent(fqg, pl), _) if bestPlan == pl => fqg
     }.get
-    val (p1, p2) = connectedPlans(PlannedComponent(bestQG, bestPlan))
+    val (p1, p2) = joins(PlannedComponent(bestQG, bestPlan))
 
     plans - p1 - p2 + PlannedComponent(bestQG, bestPlan)
   }
 
+  /**
+    * Used to plan a large amount of query parts together. Produces a left deep tree sorted by the cost of the query parts.
+    */
```

Comment:
`Used to plan a large amount ...` sounds like the comment is describing the history of the function. I suggest:

`Plans a large amount ...`